numba_dpex/ocl/mathdecl.py

```diff
@@ -230,10 +230,24 @@ class Math_atanh(Math_unary):
 
 class Math_floor(Math_unary):
     key = math.floor
+    cases = [
+        signature(types.intp, types.intp),
+        signature(types.int64, types.int64),
+        signature(types.uint64, types.uint64),
+        signature(types.int64, types.float32),
+        signature(types.int64, types.float64),
+    ]
 
 
 class Math_ceil(Math_unary):
     key = math.ceil
+    cases = [
+        signature(types.intp, types.intp),
```

Comment:
This fix is suspicious.
Original `OpenCL` function has `float (float)` signature. But we are telling Numba that it is `int (float)` signature.
If this really works that means, that `llvm` (we are providing `float (float)` siganture to `llvm`) introduce somewhere cast. It is better to either to add a comment why this is working at all, or to reimplement it in Numba adding explicit cast.

Also for `int` input we could do nothing, but instead we would cast `int` to `float`, then call `ceil` and then cast back.

@diptorupd what is your opinion?

Reply:
@AlexanderKalistratov The flow is something like this (I might be missing few things): The `ocl/mathdecl.py` module registers the allowed signatures for the `math.ceil` function that are provided in the `DpexKernelTargetContext`. Then in the `ocl/mathimpl.py` module we register the actual functions that should be used by Numba when lowering a call to `math.ceil`. 

After that when Numba lowers an actual instance of the call, it uses the usual machinery inside `numba/core/cpython/mathimpl`, but looks up the registery and uses our math function than the default one. It is inside Numba that the cast is applied when it sees that a specific `math.ceil` is supposed to return an integer. 

https://github.com/numba/numba/blob/30df74943112acc2706a98630bba6ca1f310e0c8/numba/cpython/mathimpl.py#L108.